Geolocation control

revant opened the pull request:
![map_field](https://user-images.githubusercontent.com/1040161/31609412-b33efcac-b291-11e7-9f95-9140e54e0b7f.gif)

- [x] Map field based on leaflet js
- [x] Leaflet Draw plugin to draw and store geojson [featurecollection](http://leafletjs.com/examples/geojson/) 
- [x] Leaflet Locate button to locate to current geolocation
- [x] Save
- [x] Organize code
- [x] Layers from db not shown as editable, needs fixes
- [x] On create of layer, it is added twice.
- [x] Line and area colors should match ERPNext colors 
- [x] New form (ctrl + b) needs fixes
- [x] locate center of all saved features and start map there.
- [x] basic test
- [x] Mute GeoJSON dump in print format
- [x] Documentation https://github.com/frappe/erpnext/pull/11294
- [x] Rename to most appropriate name. (Geolocation)

revant committed:
added static map field

revant committed:
leaflet draw plugin added to map control

revant committed:
Editable layer working no icons

revant committed:
Save and load data in form field

revant committed:
images and icons working

revant committed:
Locate plugin added, and loads on geolocation

pratu16x7 commented:
👍 
When does the data field populate? That could be kept read only, and maybe shown on demand :)

netchampfaris commented:
@revant Should we call it Geolocation?
Also, the value that it stores should be hidden by default and viewable on action.

revant commented:
In controls/map.js `map.on()` for events 'draw:created', 'draw:edited', 'draw:deleted', I'm setting `this.set_value(stringified_geojson_featurecollection)`

I've added eveything under make_wrapper(). Trying to understand other inherited functions for control.

netchampfaris commented:
`format_for_input` should take raw value from db and render the map
`parse` should extract value from the map and build json

Editing events should pass the json value to `set_value`

aboganas commented:
I think **GeoZone**  is better than Geolocation since you are defining a region not a location.

revant committed:
organized map control code

agritheory commented:
This is awesome.

revant committed:
loads layer as per form

agritheory commented:
@revant I have a low-priority styling request: can you change the color of the polygon lines to the Frappe purple and then shade the inside with something similar but lighter.

agritheory commented:
Also, will this have a web view? That seems like another can of worms...

revant committed:
new form clears editableLayers

revant committed:
Merge branch 'develop' of frappe/frappe into map_control

revant committed:
update: leaflet 1.2.0 and leaflet-draw 0.4.2

revant committed:
changed to ERPNext colors

revant committed:
Multiple map fields and fixes

revant committed:
fixes codacy suggestions

revant committed:
Merge branch 'develop' of frappe/frappe into map_control, resolved conflict

revant committed:
loaded geojson data editable

revant committed:
Map layers editable

revant committed:
Merge branch 'develop' of frappe/frappe into map_control

revant committed:
add only one layer on draw:create

revant committed:
Merge branch 'develop' of frappe/frappe into map_control

pratu16x7 commented on frappe/public/js/frappe/form/controls/map.js:
```diff
@@ -0,0 +1,116 @@
+	format_for_input(value) {
+		this.map.removeLayer(this.editableLayers);
+		if(value){
+			this.add_non_group_layers(L.geoJson(JSON.parse(value)), this.editableLayers);
```
`set_value()` calls `format_for_input()` internally. So in effect currently, `add_non_group_layers` is called twice (and exponentially) for every layer added. You can choose to do the adding in a single place, either in the binder or in `format_for_input()`

pratu16x7 commented on frappe/public/js/frappe/form/controls/map.js:
```diff
@@ -0,0 +1,116 @@
+		// var lc = L.control.locate().addTo(this.map);
+		// lc.start();
+
+		this.map.on('draw:created', (e) => {
```
Binder methods could be in helpers as well, being called in turn from `make_wrapper()`

pratu16x7 commented on frappe/public/js/frappe/form/controls/map.js:
```diff
@@ -0,0 +1,116 @@
+		}
+	},
+
+	add_non_group_layers(source_layer, target_group) {
```
Given the naming, seems like it adds group layers as well ;) Maybe an even more atomic method that adds a single layer?

revant committed:
code organized with helper functions

pratu16x7 commented on frappe/public/js/frappe/form/controls/map.js:
```diff
@@ -0,0 +1,125 @@
+			this.set_value(JSON.stringify(layer.toGeoJSON()));
+		});
+
+		this.map.on('draw:deleted', (e) => {
```
Can you pass in the `draw:edited` event here as well? Seems both events need the same handler. Or we could have a common handler function (DRY)

revant committed:
Replaced Leaflet Draw with Leaflet Editable

revant committed:
read values from database

revant committed:
layers saved in db

revant committed:
Using Patched Leaflet Draw (mobile-friendly)

revant committed:
Working Map field with multiple forms open

revant committed:
Merge branch 'develop' of frappe/frappe into map_control

revant committed:
Leaflet Draw css image path fix

revant committed:
Leaflet Draw

Fixes Leaflet draw css image paths
Fixes Circle marker and Circle

revant committed:
locate to geolocation or set featurecollection as center

revant committed:
[fix] leaflet fitBounds padding 50,50

revant committed:
[Fix] Leaflet (auto) locate

revant committed:
Merge branch 'develop' of frappe/frappe into map_control

revant committed:
Map field basic test

revant committed:
added refresh button on map

revant commented:
Related PR

https://github.com/frappe/erpnext/pull/11276

revant committed:
Mute map geojson in print format

revant committed:
Merge branch 'develop' of frappe/frappe into map_control

revant committed:
renamed Map field to Geolocation

netchampfaris commented on frappe/model/db_schema.py:
```diff
@@ -45,6 +45,7 @@ class InvalidColumnName(frappe.ValidationError): pass
 	,'Signature':	('longtext', '')
 	,'Color':		('varchar', varchar_len)
 	,'Barcode':		('longtext', '')
+	,'Geolocation':	('longtext', '')
```
fix formatting

revant replied:
some how sublime is not showing any funny formatting?!
![screenshot from 2017-10-25 13-51-40](https://user-images.githubusercontent.com/1040161/31988128-a9d63b02-b98b-11e7-9bbd-12819376e88a.png)

I'll add extra tab to all?

netchampfaris commented on frappe/public/js/frappe/form/controls/geolocation.js:
```diff
@@ -0,0 +1,183 @@
+		let $input_wrapper = this.$wrapper.find('.control-input-wrapper');
+		this.map_area = $(
+			`<div class="map-wrapper border">
+				<div id="` + this.df.fieldname + `" style="min-height: 400px; z-index: 1; max-width:100%"></div>
```
avoid id, put styles in form.less

revant replied:
id is required for Leaflet Map
http://leafletjs.com/reference-1.2.0.html#map
```
this.map = L.map(this.df.fieldname).setView([19.0800, 72.8961], 13);
```

netchampfaris commented on frappe/public/js/frappe/form/controls/geolocation.js:
```diff
@@ -0,0 +1,183 @@
+			draw: {
+				polyline: {
+					shapeOptions: {
+						color: '#7573fc',
```
frappe.ui.color.get('blue') ?

netchampfaris reviewed (changes requested):
Let's get this merged! (Also fix codacy 😅 )

revant committed:
Suggested changes and fixes

revant committed:
Merge branch 'develop' of frappe/frappe into map_control

netchampfaris committed:
eslint disable

netchampfaris committed:
csslint allow important
